File: ios/Apprch/ViewModels/GroupStore.swift
import Foundation
import FirebaseAuth
import FirebaseFirestore

enum GroupStore {
    static func create(name: String, solo: Bool) async throws {
        if !solo {
            _ = try await ensurePersonal()
        }
        _ = try await createSpace(name: name, solo: solo, makeActive: true)
    }

    @discardableResult
    static func ensurePersonal() async throws -> String {
        guard let uid = Auth.auth().currentUser?.uid else {
            throw GroupStoreError.signedOut
        }
        let db = Firestore.firestore()
        let userRef = db.collection("users").document(uid)
        let userSnap = try await userRef.getDocument()
        let data = userSnap.data() ?? [:]

        if let personalId = data["personalGroupId"] as? String, !personalId.isEmpty {
            return personalId
        }

        if let groupId = data["groupId"] as? String,
           let group = try await db.collection("groups").document(groupId).getDocument().data(),
           group["solo"] as? Bool == true {
            try await userRef.setData(
                [
                    "personalGroupId": groupId,
                    "groupIds": FieldValue.arrayUnion([groupId]),
                    "activeGroupId": data["activeGroupId"] as? String ?? groupId,
                    "groupId": data["groupId"] as? String ?? groupId,
                    "solo": data["solo"] as? Bool ?? true
                ],
                merge: true
            )
            return groupId
        }

        return try await createSpace(name: "Solo", solo: true, makeActive: data["groupId"] == nil)
    }

    static func migrateLegacyUserIfNeeded() async throws {
        guard let uid = Auth.auth().currentUser?.uid else { return }
        let userRef = Firestore.firestore().collection("users").document(uid)
        let data = try await userRef.getDocument().data() ?? [:]
        if let ids = data["groupIds"] as? [Any], !ids.isEmpty { return }
        guard let groupId = data["groupId"] as? String else { return }

        let solo = data["solo"] as? Bool ?? false
        if solo {
            try await userRef.setData(
                [
                    "personalGroupId": groupId,
                    "groupIds": [groupId],
                    "activeGroupId": groupId
                ],
                merge: true
            )
            return
        }

        let personalId = try await ensurePersonal()
        try await userRef.setData(
            [
                "groupIds": FieldValue.arrayUnion([groupId, personalId]),
                "activeGroupId": groupId,
                "groupId": groupId,
                "solo": false
            ],
            merge: true
        )
    }

    static func join(inviteCode: String) async throws {
        guard let uid = Auth.auth().currentUser?.uid else {
            throw GroupStoreError.signedOut
        }

        _ = try await ensurePersonal()

        let code = inviteCode.trimmingCharacters(in: .whitespacesAndNewlines).uppercased()
        let db = Firestore.firestore()
        let inviteSnap = try await db.collection("inviteCodes").document(code).getDocument()
        guard let groupId = inviteSnap.data()?["groupId"] as? String else {
            throw GroupStoreError.invalidCode
        }

        let batch = db.batch()
        batch.updateData(
            ["memberUids": FieldValue.arrayUnion([uid])],
            forDocument: db.collection("groups").document(groupId)
        )
        batch.setData(
            [
                "groupIds": FieldValue.arrayUnion([groupId]),
                "groupId": groupId,
                "activeGroupId": groupId,
                "solo": false
            ],
            forDocument: db.collection("users").document(uid),
            merge: true
        )
        try await batch.commit()
    }

    static func setActiveSpace(_ groupId: String) async throws {
        guard let uid = Auth.auth().currentUser?.uid else {
            throw GroupStoreError.signedOut
        }
        let snapshot = try await Firestore.firestore().collection("groups").document(groupId).getDocument()
        guard snapshot.exists else { throw GroupStoreError.unavailable }
        let solo = snapshot.data()?["solo"] as? Bool ?? false
        try await Firestore.firestore().collection("users").document(uid).setData(
            [
                "groupId": groupId,
                "activeGroupId": groupId,
                "solo": solo
            ],
            merge: true
        )
    }

    static func directoryFields(displayName: String, email: String?) -> [String: Any] {
        let name = displayName.trimmingCharacters(in: .whitespacesAndNewlines)
        let searchName = name.lowercased()
        var tokens = Set(searchName.split(whereSeparator: \.isWhitespace).map(String.init))
        if !searchName.isEmpty { tokens.insert(searchName) }
        var fields: [String: Any] = [
            "displayName": name,
            "searchName": searchName,
            "nameTokens": Array(tokens)
        ]
        if let email, !email.isEmpty {
            fields["email"] = email.lowercased()
        }
        return fields
    }

    static func searchUsers(query: String, excluding excludedIds: Set<String>) async throws -> [SpaceMember] {
        let needle = query.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard needle.count >= 2 else { return [] }

        let db = Firestore.firestore()
        let token = needle.split(whereSeparator: \.isWhitespace).first.map(String.init) ?? needle
        async let prefixDocs = db.collection("users")
            .whereField("searchName", isGreaterThanOrEqualTo: needle)
            .whereField("searchName", isLessThan: needle + "\u{f8ff}")
            .limit(to: 12)
            .getDocuments()
        async let tokenDocs = db.collection("users")
            .whereField("nameTokens", arrayContains: token)
            .limit(to: 12)
            .getDocuments()

        var snapshots = [try await prefixDocs, try await tokenDocs]
        if needle.contains("@") {
            snapshots.append(
                try await db.collection("users")
                    .whereField("email", isEqualTo: needle)
                    .limit(to: 5)
                    .getDocuments()
            )
        }

        var seen = Set<String>()
        var matches: [SpaceMember] = []
        for snapshot in snapshots {
            for doc in snapshot.documents {
                guard !excludedIds.contains(doc.documentID), seen.insert(doc.documentID).inserted else { continue }
                let data = doc.data()
                let name = (data["displayName"] as? String)?
                    .trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty ?? "Someone"
                matches.append(
                    SpaceMember(
                        id: doc.documentID,
                        name: name,
                        photoBase64: data["photoBase64"] as? String,
                        email: data["email"] as? String
                    )
                )
            }
        }
        return matches.sorted { $0.name.localizedCaseInsensitiveCompare($1.name) == .orderedAscending }
    }

    static func addMember(_ uid: String, toGroupId groupId: String) async throws {
        guard let me = Auth.auth().currentUser?.uid else {
            throw GroupStoreError.signedOut
        }
        guard uid != me else { throw GroupStoreError.alreadyMember }

        let groupRef = Firestore.firestore().collection("groups").document(groupId)
        let snapshot = try await groupRef.getDocument()
        guard snapshot.exists, let data = snapshot.data() else {
            throw GroupStoreError.unavailable
        }
        if data["solo"] as? Bool == true {
            throw GroupStoreError.soloSpace
        }
        let members = data["memberUids"] as? [String] ?? []
        guard members.contains(me) else {
            throw GroupStoreError.unavailable
        }
        if members.contains(uid) {
            throw GroupStoreError.alreadyMember
        }
        try await groupRef.updateData(["memberUids": FieldValue.arrayUnion([uid])])
    }

    static func members(in groupId: String) async throws -> [SpaceMember] {
        let snapshot = try await Firestore.firestore().collection("groups").document(groupId).getDocument()
        let uids = snapshot.data()?["memberUids"] as? [String] ?? []
        var members: [SpaceMember] = []
        for uid in uids {
            let user = try await Firestore.firestore().collection("users").document(uid).getDocument()
            let data = user.data() ?? [:]
            members.append(
                SpaceMember(
                    id: uid,
                    name: (data["displayName"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty ?? "Someone",
                    photoBase64: data["photoBase64"] as? String
                )
            )
        }
        return members.sorted { $0.name.localizedCaseInsensitiveCompare($1.name) == .orderedAscending }
    }

    static func spaces() async throws -> [Space] {
        guard let uid = Auth.auth().currentUser?.uid else {
            throw GroupStoreError.signedOut
        }
        let snapshot = try await Firestore.firestore()
            .collection("groups")
            .whereField("memberUids", arrayContains: uid)
            .getDocuments()
        return snapshot.documents.compactMap { doc -> Space? in
            let data = doc.data()
            guard let name = data["name"] as? String else { return nil }
            return Space(id: doc.documentID, name: name, solo: data["solo"] as? Bool ?? false)
        }
        .sorted { lhs, rhs in
            if lhs.solo != rhs.solo { return lhs.solo && !rhs.solo }
            return lhs.name.localizedCaseInsensitiveCompare(rhs.name) == .orderedAscending
        }
    }

    static func inviteCode(for groupId: String) async throws -> (code: String?, solo: Bool, name: String) {
        let snapshot = try await Firestore.firestore()
            .collection("groups")
            .document(groupId)
            .getDocument()
        guard snapshot.exists, let data = snapshot.data() else {
            throw GroupStoreError.unavailable
        }
        let code = data["inviteCode"] as? String
        let trimmed = code?.trimmingCharacters(in: .whitespacesAndNewlines)
        return (
            (trimmed?.isEmpty == false) ? trimmed : nil,
            data["solo"] as? Bool ?? false,
            data["name"] as? String ?? ""
        )
    }

    @discardableResult
    static func enableInvites(groupId: String, name: String? = nil) async throws -> String {
        guard let uid = Auth.auth().currentUser?.uid else {
            throw GroupStoreError.signedOut
        }

        let db = Firestore.firestore()
        let groupRef = db.collection("groups").document(groupId)
        let snapshot = try await groupRef.getDocument()
        guard snapshot.exists, let data = snapshot.data() else {
            throw GroupStoreError.unavailable
        }
        let members = data["memberUids"] as? [String] ?? []
        guard members.contains(uid) else {
            throw GroupStoreError.unavailable
        }
        if data["solo"] as? Bool == true {
            throw GroupStoreError.soloSpace
        }

        let existing = (data["inviteCode"] as? String)?
            .trimmingCharacters(in: .whitespacesAndNewlines)
        if let existing, !existing.isEmpty {
            return existing
        }

        let code = randomInviteCode()
        let batch = db.batch()
        batch.updateData(["inviteCode": code], forDocument: groupRef)
        batch.setData(
            ["groupId": groupId, "createdByUid": uid],
            forDocument: db.collection("inviteCodes").document(code)
        )
        try await batch.commit()
        return code
    }

    static func moveTrigger(_ triggerId: String, toGroupId: String) async throws {
        guard Auth.auth().currentUser?.uid != nil else {
            throw GroupStoreError.signedOut
        }
        let db = Firestore.firestore()
        let taskRef = db.collection("tasks").document(triggerId)
        let taskSnap = try await taskRef.getDocument()
        guard taskSnap.exists else { throw GroupStoreError.unavailable }
        guard let fromGroupId = taskSnap.data()?["groupId"] as? String else {
            throw GroupStoreError.unavailable
        }
        guard fromGroupId != toGroupId else { return }

        try await taskRef.updateData(["groupId": toGroupId])

        let events = try await db.collection("events")
            .whereField("groupId", isEqualTo: fromGroupId)
            .whereField("taskId", isEqualTo: triggerId)
            .getDocuments()
        guard !events.documents.isEmpty else { return }

        let batch = db.batch()
        for doc in events.documents {
            batch.updateData(["groupId": toGroupId], forDocument: doc.reference)
        }
        try await batch.commit()
    }

    @discardableResult
    static func createGroup(named name: String, movingTriggerId triggerId: String?) async throws -> String {
        let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { throw GroupStoreError.needsName }
        _ = try await ensurePersonal()
        let groupId = try await createSpace(name: trimmed, solo: false, makeActive: true)
        if let triggerId {
            try await moveTrigger(triggerId, toGroupId: groupId)
        }
        return groupId
    }

    static func userFacingMessage(for error: Error) -> String {
        if let storeError = error as? GroupStoreError {
            return storeError.localizedDescription
        }
        let text = error.localizedDescription
        if text.localizedCaseInsensitiveContains("not found")
            || text.localizedCaseInsensitiveContains("permission") {
            return "Couldn’t finish that. Check your connection and try again."
        }
        return text
    }

    @discardableResult
    private static func createSpace(name: String, solo: Bool, makeActive: Bool) async throws -> String {
        guard let uid = Auth.auth().currentUser?.uid else {
            throw GroupStoreError.signedOut
        }

        let db = Firestore.firestore()
        let groupRef = db.collection("groups").document()
        let inviteCode = solo ? nil : randomInviteCode()
        var group: [String: Any] = [
            "name": name,
            "solo": solo,
            "memberUids": [uid],
            "createdAt": FieldValue.serverTimestamp()
        ]
        if let inviteCode {
            group["inviteCode"] = inviteCode
        }

        var userUpdate: [String: Any] = [
            "groupIds": FieldValue.arrayUnion([groupRef.documentID])
        ]
        if solo {
            userUpdate["personalGroupId"] = groupRef.documentID
        }
        if makeActive {
            userUpdate["groupId"] = groupRef.documentID
            userUpdate["activeGroupId"] = groupRef.documentID
            userUpdate["solo"] = solo
        }

        let batch = db.batch()
        batch.setData(group, forDocument: groupRef)
        batch.setData(userUpdate, forDocument: db.collection("users").document(uid), merge: true)
        if let inviteCode {
            batch.setData(
                ["groupId": groupRef.documentID, "createdByUid": uid],
                forDocument: db.collection("inviteCodes").document(inviteCode)
            )
        }
        try await batch.commit()
        return groupRef.documentID
    }

    private static func randomInviteCode() -> String {
        let chars = Array("ABCDEFGHJKLMNPQRSTUVWXYZ23456789")
        return String((0..<6).map { _ in chars.randomElement()! })
    }
}

enum GroupStoreError: LocalizedError {
    case signedOut
    case invalidCode
    case soloSpace
    case unavailable
    case needsName
    case alreadyMember

    var errorDescription: String? {
        switch self {
        case .signedOut: return "Please sign in again."
        case .invalidCode: return "That invite code isn’t valid."
        case .soloSpace: return "Invite from a group, or move a Task into a new group."
        case .unavailable: return "Couldn’t load this space."
        case .needsName: return "Give the group a name first."
        case .alreadyMember: return "They’re already in this group."
        }
    }
}

File: android/app/src/main/java/com/apprch/app/data/GroupStore.kt
package com.apprch.app.data

import com.apprch.app.model.Space
import com.apprch.app.model.SpaceMember
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.SetOptions
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.tasks.await

object GroupStore {
    private val inviteChars = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789"

    fun directoryFields(displayName: String, email: String?): Map<String, Any> {
        val name = displayName.trim()
        val searchName = name.lowercase()
        val tokens = searchName.split(Regex("\\s+")).filter { it.isNotEmpty() }.toMutableSet()
        if (searchName.isNotEmpty()) tokens += searchName
        val fields = hashMapOf<String, Any>(
            "displayName" to name,
            "searchName" to searchName,
            "nameTokens" to tokens.toList()
        )
        if (!email.isNullOrBlank()) {
            fields["email"] = email.lowercase()
        }
        return fields
    }

    suspend fun create(name: String, solo: Boolean) {
        if (!solo) ensurePersonal()
        createSpace(name = name, solo = solo, makeActive = true)
    }

    suspend fun ensurePersonal(): String {
        val uid = FirebaseAuth.getInstance().currentUser?.uid
            ?: throw IllegalStateException("Please sign in again.")
        val db = FirebaseFirestore.getInstance()
        val userRef = db.collection("users").document(uid)
        val data = userRef.get().await().data ?: emptyMap()
        val personalId = data["personalGroupId"] as? String
        if (!personalId.isNullOrEmpty()) return personalId

        val groupId = data["groupId"] as? String
        if (groupId != null) {
            val group = db.collection("groups").document(groupId).get().await().data
            if (group?.get("solo") == true) {
                userRef.set(
                    mapOf(
                        "personalGroupId" to groupId,
                        "groupIds" to FieldValue.arrayUnion(groupId),
                        "activeGroupId" to (data["activeGroupId"] ?: groupId),
                        "groupId" to (data["groupId"] ?: groupId),
                        "solo" to (data["solo"] ?: true)
                    ),
                    SetOptions.merge()
                ).await()
                return groupId
            }
        }
        return createSpace(name = "Solo", solo = true, makeActive = data["groupId"] == null)
    }

    suspend fun join(inviteCode: String) {
        val uid = FirebaseAuth.getInstance().currentUser?.uid
            ?: throw IllegalStateException("Please sign in again.")
        ensurePersonal()
        val code = inviteCode.trim().uppercase()
        val db = FirebaseFirestore.getInstance()
        val invite = db.collection("inviteCodes").document(code).get().await()
        val groupId = invite.getString("groupId")
            ?: throw IllegalStateException("That invite code isn’t valid.")

        val batch = db.batch()
        batch.update(
            db.collection("groups").document(groupId),
            "memberUids",
            FieldValue.arrayUnion(uid)
        )
        batch.set(
            db.collection("users").document(uid),
            mapOf(
                "groupIds" to FieldValue.arrayUnion(groupId),
                "groupId" to groupId,
                "activeGroupId" to groupId,
                "solo" to false
            ),
            SetOptions.merge()
        )
        batch.commit().await()
    }

    suspend fun setActiveSpace(groupId: String) {
        val uid = FirebaseAuth.getInstance().currentUser?.uid
            ?: throw IllegalStateException("Please sign in again.")
        val snapshot = FirebaseFirestore.getInstance().collection("groups").document(groupId).get().await()
        if (!snapshot.exists()) throw IllegalStateException("Couldn’t load this space.")
        val solo = snapshot.getBoolean("solo") ?: false
        FirebaseFirestore.getInstance().collection("users").document(uid).set(
            mapOf(
                "groupId" to groupId,
                "activeGroupId" to groupId,
                "solo" to solo
            ),
            SetOptions.merge()
        ).await()
    }

    suspend fun spaces(): List<Space> {
        val uid = FirebaseAuth.getInstance().currentUser?.uid
            ?: throw IllegalStateException("Please sign in again.")
        val snapshot = FirebaseFirestore.getInstance()
            .collection("groups")
            .whereArrayContains("memberUids", uid)
            .get()
            .await()
        return snapshot.documents.mapNotNull { doc ->
            val name = doc.getString("name") ?: return@mapNotNull null
            Space(id = doc.id, name = name, solo = doc.getBoolean("solo") ?: false)
        }.sortedWith(compareBy<Space> { !it.solo }.thenBy { it.name.lowercase() })
    }

    suspend fun members(groupId: String): List<SpaceMember> {
        val db = FirebaseFirestore.getInstance()
        val snapshot = db.collection("groups").document(groupId).get().await()
        val uids = snapshot.get("memberUids") as? List<*> ?: emptyList<Any>()
        return uids.mapNotNull { it as? String }.map { uid ->
            val data = db.collection("users").document(uid).get().await().data ?: emptyMap()
            SpaceMember(
                id = uid,
                name = (data["displayName"] as? String)?.trim()?.ifEmpty { null } ?: "Someone",
                photoBase64 = data["photoBase64"] as? String,
                email = data["email"] as? String
            )
        }.sortedBy { it.name.lowercase() }
    }

    suspend fun searchUsers(query: String, excluding: Set<String>): List<SpaceMember> {
        val needle = query.trim().lowercase()
        if (needle.length < 2) return emptyList()
        val db = FirebaseFirestore.getInstance()
        val token = needle.split(Regex("\\s+")).first()
        return coroutineScope {
            val prefix = async {
                db.collection("users")
                    .whereGreaterThanOrEqualTo("searchName", needle)
                    .whereLessThan("searchName", needle + "\uf8ff")
                    .limit(12)
                    .get()
                    .await()
            }
            val tokens = async {
                db.collection("users")
                    .whereArrayContains("nameTokens", token)
                    .limit(12)
                    .get()
                    .await()
            }
            val docs = (prefix.await().documents + tokens.await().documents).toMutableList()
            if (needle.contains("@")) {
                docs += db.collection("users")
                    .whereEqualTo("email", needle)
                    .limit(5)
                    .get()
                    .await()
                    .documents
            }
            val seen = mutableSetOf<String>()
            docs.mapNotNull { doc ->
                if (doc.id in excluding || !seen.add(doc.id)) return@mapNotNull null
                val data = doc.data ?: return@mapNotNull null
                SpaceMember(
                    id = doc.id,
                    name = (data["displayName"] as? String)?.trim()?.ifEmpty { null } ?: "Someone",
                    photoBase64 = data["photoBase64"] as? String,
                    email = data["email"] as? String
                )
            }.sortedBy { it.name.lowercase() }
        }
    }

    suspend fun addMember(uid: String, groupId: String) {
        val me = FirebaseAuth.getInstance().currentUser?.uid
            ?: throw IllegalStateException("Please sign in again.")
        if (uid == me) throw IllegalStateException("They’re already in this group.")
        val groupRef = FirebaseFirestore.getInstance().collection("groups").document(groupId)
        val snapshot = groupRef.get().await()
        val data = snapshot.data ?: throw IllegalStateException("Couldn’t load this space.")
        if (data["solo"] == true) {
            throw IllegalStateException("Invite from a group, or move a Task into a new group.")
        }
        val members = data["memberUids"] as? List<*> ?: emptyList<Any>()
        if (!members.contains(me)) throw IllegalStateException("Couldn’t load this space.")
        if (members.contains(uid)) throw IllegalStateException("They’re already in this group.")
        groupRef.update("memberUids", FieldValue.arrayUnion(uid)).await()
    }

    suspend fun enableInvites(groupId: String): String {
        val uid = FirebaseAuth.getInstance().currentUser?.uid
            ?: throw IllegalStateException("Please sign in again.")
        val db = FirebaseFirestore.getInstance()
        val groupRef = db.collection("groups").document(groupId)
        val snapshot = groupRef.get().await()
        val data = snapshot.data ?: throw IllegalStateException("Couldn’t load this group.")
        val members = data["memberUids"] as? List<*> ?: emptyList<Any>()
        if (!members.contains(uid)) throw IllegalStateException("Couldn’t load this group.")
        if (data["solo"] as? Boolean == true) {
            throw IllegalStateException("Invite from a group, or move a Task into a new group.")
        }
        val existing = (data["inviteCode"] as? String)?.trim()
        if (!existing.isNullOrEmpty()) return existing

        val code = randomInviteCode()
        val batch = db.batch()
        batch.update(groupRef, mapOf("inviteCode" to code))
        batch.set(
            db.collection("inviteCodes").document(code),
            mapOf("groupId" to groupId, "createdByUid" to uid)
        )
        batch.commit().await()
        return code
    }

    suspend fun inviteCode(groupId: String): Triple<String?, Boolean, String> {
        val snapshot = FirebaseFirestore.getInstance().collection("groups").document(groupId).get().await()
        val data = snapshot.data ?: throw IllegalStateException("Couldn’t load this space.")
        val code = (data["inviteCode"] as? String)?.trim()?.ifEmpty { null }
        return Triple(code, data["solo"] as? Boolean ?: false, data["name"] as? String ?: "")
    }

    suspend fun moveTask(taskId: String, toGroupId: String) {
        FirebaseAuth.getInstance().currentUser
            ?: throw IllegalStateException("Please sign in again.")
        val db = FirebaseFirestore.getInstance()
        val taskRef = db.collection("tasks").document(taskId)
        val task = taskRef.get().await()
        if (!task.exists()) throw IllegalStateException("Couldn’t load this space.")
        val fromGroupId = task.getString("groupId")
            ?: throw IllegalStateException("Couldn’t load this space.")
        if (fromGroupId == toGroupId) return
        taskRef.update("groupId", toGroupId).await()

        val events = db.collection("events")
            .whereEqualTo("groupId", fromGroupId)
            .whereEqualTo("taskId", taskId)
            .get()
            .await()
        if (events.isEmpty) return
        val batch = db.batch()
        events.documents.forEach { batch.update(it.reference, "groupId", toGroupId) }
        batch.commit().await()
    }

    suspend fun createGroup(named: String, movingTaskId: String?): String {
        val trimmed = named.trim()
        if (trimmed.isEmpty()) throw IllegalStateException("Give the group a name first.")
        ensurePersonal()
        val groupId = createSpace(name = trimmed, solo = false, makeActive = true)
        if (movingTaskId != null) {
            moveTask(movingTaskId, groupId)
        }
        return groupId
    }

    fun userFacingMessage(error: Throwable): String {
        val text = error.localizedMessage.orEmpty()
        if (text.contains("not found", ignoreCase = true) ||
            text.contains("permission", ignoreCase = true)
        ) {
            return "Couldn’t finish that. Check your connection and try again."
        }
        return text.ifBlank { "Couldn’t finish that. Try again." }
    }

    private suspend fun createSpace(name: String, solo: Boolean, makeActive: Boolean): String {
        val uid = FirebaseAuth.getInstance().currentUser?.uid
            ?: throw IllegalStateException("Please sign in again.")
        val db = FirebaseFirestore.getInstance()
        val groupRef = db.collection("groups").document()
        val inviteCode = if (solo) null else randomInviteCode()
        val group = hashMapOf<String, Any>(
            "name" to name,
            "solo" to solo,
            "memberUids" to listOf(uid),
            "createdAt" to FieldValue.serverTimestamp()
        )
        if (inviteCode != null) group["inviteCode"] = inviteCode

        val userUpdate = hashMapOf<String, Any>(
            "groupIds" to FieldValue.arrayUnion(groupRef.id)
        )
        if (solo) userUpdate["personalGroupId"] = groupRef.id
        if (makeActive) {
            userUpdate["groupId"] = groupRef.id
            userUpdate["activeGroupId"] = groupRef.id
            userUpdate["solo"] = solo
        }

        val batch = db.batch()
        batch.set(groupRef, group)
        batch.set(db.collection("users").document(uid), userUpdate, SetOptions.merge())
        if (inviteCode != null) {
            batch.set(
                db.collection("inviteCodes").document(inviteCode),
                mapOf("groupId" to groupRef.id, "createdByUid" to uid)
            )
        }
        batch.commit().await()
        return groupRef.id
    }

    private fun randomInviteCode(): String =
        (1..6).map { inviteChars.random() }.joinToString("")
}
